source/OrderedTypes/InductiveTypesfromSmallGeneratedLattice.lagda

```diff
@@ -0,0 +1,2209 @@
+         sup-𝓘-≤-sup-P : (sup-𝓘 ≤ sup-P) holds
+         sup-𝓘-≤-sup-P = joins-preserve-containment
+                          {𝓘'-subset}
+                          {P-a}
+                          𝓘'-subset-⊆-P-a
+         sup-P-＝-a : sup-P ＝ a
+         sup-P-＝-a = is-sup'ᴮ a ⁻¹
+
+\end{code}
+
+We now consider a boundedness restricion on inductive definitions and show
+that bounded inductive definitions are local.
+
+An inductive definition is bounded if there is a small indexed family of
+types that can be substituted for elements a : L in a sense to be made
+precise below.
+
+\begin{code}
+
+module bounded-inductive-definition {𝓤 𝓦 𝓥 : Universe}
+                                    {B : 𝓥  ̇}
+                                    (L : Sup-Lattice 𝓤 𝓦 𝓥)
+                                    (q : B → ⟨ L ⟩)
+                                     where
+
+ open small-basis L q
+ open Joins _≤_
+
+ module bounded-from-small-basis-facts (h : is-small-basis) where
+
+  open small-basis-facts h
+  open PropositionalTruncation pt
+
+  _is-a-small-cover-of_ : (X : 𝓥  ̇) → (Y : 𝓣  ̇) → 𝓥 ⊔ 𝓣  ̇
+  X is-a-small-cover-of Y = X ↠ Y
+
+  _has-a-bound : (ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)) → 𝓤 ⊔ 𝓦 ⊔ (𝓥 ⁺)  ̇
+  ϕ has-a-bound = Σ I ꞉ 𝓥  ̇ , Σ α ꞉ (I → 𝓥  ̇) ,
+                   ((a : ⟨ L ⟩)
+                 → (b : B)
+                 → (b , a) ∈ ϕ
+                 → (Ǝ i ꞉ I , α i is-a-small-cover-of ↓ᴮ a) holds)
+
+  bound-index : {ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)} → ϕ has-a-bound → 𝓥  ̇
+  bound-index (I , α , covering) = I
+
+  bound-family : {ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)}
+               → (bnd : ϕ has-a-bound)
+               → (bound-index {ϕ} bnd → 𝓥  ̇)
+  bound-family (I , α , covering) = α
+
+  covering-condition : {ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)}
+                     → (bnd : ϕ has-a-bound)
+                     → ((a : ⟨ L ⟩)
+                     → (b : B)
+                     → (b , a) ∈ ϕ
+                     → (Ǝ i ꞉ (bound-index {ϕ} bnd) ,
+                        ((bound-family {ϕ} bnd) i is-a-small-cover-of ↓ᴮ a))
+                         holds)
+  covering-condition (I , α , covering) = covering
+
+  _is-bounded : (ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)) → 𝓤 ⊔ 𝓦 ⊔ (𝓥 ⁺)  ̇
+  ϕ is-bounded = ((a : ⟨ L ⟩) → (b : B) → ((b , a) ∈ ϕ) is 𝓥 small)
+               × (ϕ has-a-bound)
+
+  open local-inductive-definitions L q
+  open local-from-small-basis-facts h
+
+  _bounded-implies-local : (ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩))
+                         → ϕ is-bounded
+                         → ϕ is-local
+  (ϕ bounded-implies-local) (ϕ-small , ϕ-has-bound) a =
+    smallness-closed-under-≃ S₀-is-small S₀-≃-S
+   where
+    I : 𝓥  ̇
+    I = bound-index {ϕ} ϕ-has-bound
+    α : I → 𝓥  ̇
+    α = bound-family {ϕ} ϕ-has-bound
+    cov : (a : ⟨ L ⟩)
+        → (b : B)
+        → (b , a) ∈ ϕ
+        → (Ǝ i ꞉ I , (α i is-a-small-cover-of ↓ᴮ a)) holds
+    cov = covering-condition {ϕ} ϕ-has-bound
+    S₀ : 𝓤 ⊔ 𝓦 ⊔ 𝓥  ̇
+    S₀ = Σ b ꞉ B , (Ǝ i ꞉ I , (Ǝ m ꞉ (α i → ↓ᴮ a) ,
```

Comment:
This definition is awkward to work with because of the nested truncations. Notice that `Ǝ ... Ǝ  ...` is equivalent to `Ǝ ... Σ ...` and the latter will be easier to work with.